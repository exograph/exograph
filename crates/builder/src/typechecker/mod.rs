use std::collections::HashMap;

use codemap::Span;
use codemap_diagnostic::{Diagnostic, Level, SpanLabel, SpanStyle};
use core_model::{mapped_arena::MappedArena, primitive_type::PrimitiveType};
use core_model_builder::{
    ast::ast_types::{AstModel, AstModelKind, AstService, AstSystem, Untyped},
    typechecker::{
        annotation::{AnnotationSpec, AnnotationTarget, MappedAnnotationParamSpec},
        typ::{Service, Type, TypecheckedSystem},
        Scope,
    },
};
use core_plugin_interface::interface::SubsystemBuilder;

use crate::error::ParserError;

mod annotation;
pub mod annotation_map;
mod annotation_params;
mod expression;
mod field;
mod field_default_value;
mod field_type;
mod logical_op;
mod model;
mod relational_op;
mod selection;
mod service;

pub trait TypecheckInto<T> {
    fn shallow(&self) -> T;
    fn pass(
        &self,
        typ: &mut T,
        type_env: &MappedArena<Type>,
        annotation_env: &HashMap<String, AnnotationSpec>,
        scope: &Scope,
        errors: &mut Vec<Diagnostic>,
    ) -> bool;
}

pub trait TypecheckFrom<T>
where
    Self: Sized,
{
    fn shallow(untyped: &T) -> Self;
    fn pass(
        &mut self,
        type_env: &MappedArena<Type>,
        annotation_env: &HashMap<String, AnnotationSpec>,
        scope: &Scope,
        errors: &mut Vec<Diagnostic>,
    ) -> bool;
}

fn populate_type_env(env: &mut MappedArena<Type>) {
    // TODO: maybe we don't need to do this manually
    env.add("Boolean", Type::Primitive(PrimitiveType::Boolean));
    env.add("Int", Type::Primitive(PrimitiveType::Int));
    env.add("Float", Type::Primitive(PrimitiveType::Float));
    env.add("Decimal", Type::Primitive(PrimitiveType::Decimal));
    env.add("String", Type::Primitive(PrimitiveType::String));
    env.add("LocalTime", Type::Primitive(PrimitiveType::LocalTime));
    env.add(
        "LocalDateTime",
        Type::Primitive(PrimitiveType::LocalDateTime),
    );
    env.add("LocalDate", Type::Primitive(PrimitiveType::LocalDate));
    env.add("Instant", Type::Primitive(PrimitiveType::Instant));
    env.add("Json", Type::Primitive(PrimitiveType::Json));
    env.add("Blob", Type::Primitive(PrimitiveType::Blob));
    env.add("Uuid", Type::Primitive(PrimitiveType::Uuid));

    env.add("Claytip", Type::Primitive(PrimitiveType::Claytip));
    env.add("ClaytipPriv", Type::Primitive(PrimitiveType::ClaytipPriv));

    env.add(
        "Operation",
        Type::Primitive(PrimitiveType::Interception("Operation".to_string())),
    );
}

fn populate_annotation_env(
    subsystem_builders: &[Box<dyn SubsystemBuilder>],
    env: &mut HashMap<String, AnnotationSpec>,
) {
    let mut annotations = vec![
        (
            "access",
            AnnotationSpec {
                targets: &[
                    AnnotationTarget::Type,
                    AnnotationTarget::Field,
                    AnnotationTarget::Method,
                ],
                no_params: false,
                single_params: true,
                mapped_params: Some(&[
                    MappedAnnotationParamSpec {
                        name: "query",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "mutation",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "create",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "update",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "delete",
                        optional: true,
                    },
                ]),
            },
        ),
        (
            "cookie",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "env",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        #[cfg(feature = "test-context")]
        (
            "test",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "before",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "after",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "around",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "header",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "inject",
            AnnotationSpec {
                targets: &[AnnotationTarget::Argument],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "clientIp",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: false,
                mapped_params: None,
            },
        ),
        (
            "jwt",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "query",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
    ];

    for builder in subsystem_builders.iter() {
        annotations.extend(builder.annotations())
    }

    for (name, spec) in annotations {
        env.insert(name.to_owned(), spec);
    }
}

pub fn build(
    subsystem_builders: &[Box<dyn SubsystemBuilder>],
    ast_system: AstSystem<Untyped>,
) -> Result<TypecheckedSystem, ParserError> {
    let mut types_arena: MappedArena<Type> = MappedArena::default();
    let mut services_arena: MappedArena<Service> = MappedArena::default();
    let mut annotation_env = HashMap::new();
    populate_type_env(&mut types_arena);
    populate_annotation_env(subsystem_builders, &mut annotation_env);

    validate_no_duplicates(&ast_system.services, |s| &s.name, |s| s.span, "service")?;

    let mut ast_service_types: Vec<AstModel<Untyped>> = vec![];
    for service in ast_system.services.iter() {
        ast_service_types.extend(service.types.clone());
        services_arena.add(&service.name, Service(AstService::shallow(service)));

        validate_service(service)?;
    }

    let ast_types_iter = ast_system.types.iter().chain(ast_service_types.iter());
    let ast_root_types = &ast_system.types;

    for ast_type in ast_types_iter.clone() {
        types_arena.add(
            ast_type.name.as_str(),
            Type::Composite(AstModel::shallow(ast_type)),
        );
    }

    loop {
        let mut did_change = false;
        let init_scope = Scope {
            enclosing_type: None,
        };

        let mut errors = Vec::new();

        for ast_root_type in ast_root_types.iter() {
            let mut typ = types_arena
                .get_by_key(ast_root_type.name.as_str())
                .unwrap()
                .clone();
            if let Type::Composite(c) = &mut typ {
                if c.kind != AstModelKind::Context {
                    errors.push(Diagnostic {
                        level: Level::Error,
                        message: "Models and types are not permitted outside a service".to_string(),
                        code: Some("C000".to_string()),
                        spans: vec![SpanLabel {
                            span: ast_root_type.span,
                            style: SpanStyle::Primary,
                            label: None,
                        }],
                    })
                }
            }
        }

        // Temporary workaround to avoid reporting errors twice (once here, and once in the service pass)
        // TODO: Remove this after fixing https://github.com/payalabs/payas/issues/596
        let mut ignore_errors = Vec::new();

        for ast_type in ast_types_iter.clone() {
            let mut typ = types_arena
                .get_by_key(ast_type.name.as_str())
                .unwrap()
                .clone();
            if let Type::Composite(c) = &mut typ {
                let pass_res = c.pass(
                    &types_arena,
                    &annotation_env,
                    &init_scope,
                    &mut ignore_errors,
                );
                if pass_res {
                    *types_arena.get_by_key_mut(ast_type.name.as_str()).unwrap() = typ;
                    did_change = true;
                }
            } else {
                errors.push(Diagnostic {
                    level: Level::Error,
                    message: format!("Type {} is not a model", typ),
                    code: Some("C000".to_string()),
                    spans: vec![],
                });
            }
        }

        for ast_service in ast_system.services.iter() {
            let mut service = services_arena
                .get_by_key(ast_service.name.as_str())
                .unwrap()
                .clone();
            let Service(s) = &mut service;

            let pass_res = s.pass(&types_arena, &annotation_env, &init_scope, &mut errors);
            if pass_res {
                *services_arena
                    .get_by_key_mut(ast_service.name.as_str())
                    .unwrap() = service;
                did_change = true;
            }
        }

        if !did_change {
            if !errors.is_empty() {
                return Err(ParserError::Diagnosis(errors));
            } else {
                return Ok(TypecheckedSystem {
                    types: types_arena,
                    services: services_arena,
                });
            }
        }
    }
}

fn validate_service(service: &AstService<Untyped>) -> Result<(), ParserError> {
    validate_no_duplicates(
        &service.methods,
        |method| &method.name,
        |method| method.span,
        "operation",
    )?;

    validate_no_duplicates(
        &service.types,
        |model| &model.name,
        |model| model.span,
        "model/type",
    )
}

fn validate_no_duplicates<T>(
    items: &[T],
    get_name: impl Fn(&T) -> &str,
    get_span: impl Fn(&T) -> Span,
    item_kind: &str, // To print as a part of the error message
) -> Result<(), ParserError> {
    let mut items_with_pos = HashMap::new();
    let mut duplicates_with_pos = vec![];

    for item in items.iter() {
        // TODO: Use try_insert when it's stable. This way the first item will always be designated as the existing one
        // Currently, if we have three duplicates, we get diagnostics as (existing: 1, duplicate: 2), (existing: 2, duplicate: 3)
        let existing_item = items_with_pos.insert(get_name(item), get_span(item));

        if let Some(existing_span) = existing_item {
            duplicates_with_pos.push((get_name(item), existing_span, get_span(item)));
        }
    }

    if duplicates_with_pos.is_empty() {
        Ok(())
    } else {
        let diagnostics =
            duplicates_with_pos
                .into_iter()
                .map(|(name, existing_span, duplicate_span)| Diagnostic {
                    level: Level::Error,
                    message: format!("Duplicate {item_kind}: {name}"),
                    code: Some("C000".to_string()),
                    spans: vec![
                        SpanLabel {
                            span: existing_span,
                            style: SpanStyle::Primary,
                            label: Some("first defined here".to_string()),
                        },
                        SpanLabel {
                            span: duplicate_span,
                            style: SpanStyle::Secondary,
                            label: Some("again defined here".to_string()),
                        },
                    ],
                });

        Err(ParserError::Diagnosis(diagnostics.collect()))
    }
}
#[cfg(test)]
pub mod test_support {
    use codemap::CodeMap;

    use super::*;
    use crate::{load_subsystem_builders, parser::parse_str};

    pub fn build(src: &str) -> Result<TypecheckedSystem, ParserError> {
        let mut codemap = CodeMap::new();
        let parsed = parse_str(src, &mut codemap, "input.clay")?;
        let subsystem_builders = load_subsystem_builders().unwrap();
        super::build(&subsystem_builders, parsed)
    }

    pub fn parse_sorted(src: &str) -> Vec<(String, Type)> {
        let checked = build(src).unwrap();

        let mut entries: Vec<_> = checked
            .types
            .keys()
            .map(|key| (key.clone(), checked.types.get_by_key(key).unwrap().clone()))
            .collect();

        entries.sort_by_key(|pair| pair.0.to_owned());
        entries
    }
}

#[cfg(test)]
mod tests {
    use super::test_support::{build, parse_sorted};

    // Due to a change in insta version 1.12, test names (hence the snapshot names) get derived
    // from the surrounding function, so we must use a macro instead of a helper function.
    macro_rules! assert_typechecking {
        ($src:expr) => {
            insta::with_settings!({sort_maps => true}, {
                insta::assert_yaml_snapshot!(build($src).unwrap())
            });
        };
    }

    #[test]
    fn simple() {
        let src = r#"
        @postgres
        service UserService {
            type User {
              @column("custom_column") @access(self.role == "role_admin" || self.role == "role_superuser" || self.doc.is_public)
              doc: Doc;
              role: String
            }

            type Doc {
              is_public: Boolean
            }
        }
        "#;

        assert_typechecking!(src);
    }

    #[test]
    fn with_auth_context_use_in_field_annotation() {
        let src = r#"
        context AuthContext {
            @jwt role: String
        }

        @postgres
        service DocumentService {
            type Doc {
              is_public: Boolean
              @access(AuthContext.role == "ROLE_ADMIN" || self.is_public) content: String 
            }
        }
        "#;

        assert_typechecking!(src);
    }

    #[test]
    fn with_array_in_operator() {
        let src = r#"
        context AuthContext {
            @jwt roles: Array<String> 
        }

        @postgres
        service DocumentService {
            type Doc {
              @access("ROLE_ADMIN" in AuthContext.roles) content: String 
            }
        }
        "#;

        assert_typechecking!(src);
    }

    #[test]
    fn with_auth_context_use_in_type_annotation() {
        let src = r#"
        context AuthContext {
            @jwt role: String
        }
        
        @postgres
        service DocumentService {
            @access(AuthContext.role == "ROLE_ADMIN" || self.is_public)
            type Doc {
              is_public: Boolean
              content: String
            }
        }
        "#;

        assert_typechecking!(src);
    }

    #[test]
    fn insignificant_whitespace() {
        let typical = r#"
        @postgres
        service DocumentService {
            @table("venues")
            type Venue {
                @column("idx") @pk id: Int 
                name: String
            }
        }
        "#;

        let with_whitespace = r#"
        @postgres
        service      DocumentService{
        @table ( "venues" )
        type    Venue
        {
            @column(  "idx"  )    @pk  id:   Int   
           

            name:String

        }}

        "#;

        let typical_parsed = serde_json::to_string(&parse_sorted(typical)).unwrap();
        let with_whitespace_parsed = serde_json::to_string(&parse_sorted(with_whitespace)).unwrap();
        assert_eq!(typical_parsed, with_whitespace_parsed);
    }

    #[test]
    fn unknown_annotation() {
        let src = r#"
        @postgres
        service UserService {
            @asdf
            type User {
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn duplicate_annotation() {
        let src = r#"
        @postgres
        service UserService {
            @table("users")
            @table("users")
            type User {
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn duplicate_plugin_annotations() {
        let src = r#"
        @postgres
        @postgres
        service UserService {
            @table("users")
            type User {
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn no_plugin_annotation() {
        let src = r#"
        service UserService {
            type User {
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn type_at_root() {
        let src_model = r#"
        type User {
        }
        "#;

        assert_err(src_model);
    }

    #[test]
    fn context_in_a_service() {
        let src_model = r#"
        @postgres
        service UserService {
            context AuthContext {
                @jwt role: String
            }
        }
        "#;

        assert_err(src_model);
    }

    #[test]
    fn invalid_annotation_parameter_type() {
        let expected_none = r#"
        @postgres
        service UserService {
            type User {
                @pk("asdf") id: Int 
            }
        }
        "#;

        let expected_single = r#"
        @postgres
        service UserService {
            @table
            type User {
            }
        }
        "#;

        let expected_map = r#"
        @postgres
        service UserService {
            type User {
                @range(5) id: Int
            }
        }
        "#;

        assert_err(expected_none);
        assert_err(expected_single);
        assert_err(expected_map);
    }

    #[test]
    fn duplicate_annotation_mapped_param() {
        let src = r#"
        @postgres
        service UserService {
            type User {
                @range(min=5, max=10, min=3) id: Int
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn unknown_annotation_mapped_param() {
        let src = r#"
        @postgres
        service UserService {
            type User {
                @range(min=5, maxx=10) id: Int
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn invalid_annotation_target() {
        let model = r#"
        @postgres
        service UserService {
            @pk
            type User {
            }
        }
        "#;

        let field = r#"
        @postgres
        service UserService {
            type User {
                @table("asdf") id: Int
            }
        }
        "#;

        assert_err(model);
        assert_err(field);
    }

    #[test]
    fn multiple_types() {
        let model = r#"
        @deno("test.ts")
        service User {
            type User {
                id: Int
            }

            type User {
                id: Int
            }

            query userName(id: Int): String
        }
        "#;

        assert_err(model);
    }

    #[test]
    fn multiple_same_named_services() {
        let model = r#"
        @deno("foo.js")
        service Foo {
            query userName(id: Int): String
        }

        @deno("foo.js")
        service Foo {
            query userName(id: Int): String
        }
        "#;

        assert_err(model);
    }

    #[test]
    fn multiple_same_named_operations() {
        let model = r#"
        @deno("foo.js")
        service Foo {
            query userName(id1: Int): String
            query userName(id2: Int): String
            query userName(id3: Int): String
        }
        "#;

        assert_err(model);
    }

    #[test]
    fn multiple_same_named_typess() {
        let model = r#"
        @deno("foo.js")
        service Foo {
            type User {
                id: Int
                name: String
            }
            type User {
                id: Int
                name: String
            }
        }
        "#;

        assert_err(model);
    }

    #[test]
    fn multiple_same_named_types() {
        let model = r#"
        @postgres
        service Foo {
            type User {
                id: Int
                name: String
            }
            type User {
                id: Int
                name: String
            }
        }
        "#;

        assert_err(model);
    }

    #[test]
    fn multiple_same_named_model_and_types() {
        let model = r#"
        @deno("foo.js")
        service Foo {
            type User {
                id: Int
                name: String
            }
            type User {
                id: Int
                name: String
            }
        }
        "#;

        assert_err(model);
    }

    fn assert_err(src: &str) {
        assert!(build(src).is_err());
    }
}
