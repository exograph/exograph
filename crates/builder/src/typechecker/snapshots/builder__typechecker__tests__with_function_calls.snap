---
source: crates/builder/src/typechecker/mod.rs
expression: build(src).unwrap()
---
types:
  values:
    - - ~
      - Primitive: Boolean
    - - ~
      - Primitive: Int
    - - ~
      - Primitive: Float
    - - ~
      - Primitive: Decimal
    - - ~
      - Primitive: String
    - - ~
      - Primitive: LocalTime
    - - ~
      - Primitive: LocalDateTime
    - - ~
      - Primitive: LocalDate
    - - ~
      - Primitive: Instant
    - - ~
      - Primitive: Json
    - - ~
      - Primitive: Blob
    - - ~
      - Primitive: Uuid
    - - ~
      - Primitive: Vector
    - - ~
      - Primitive: Exograph
    - - ~
      - Primitive: ExographPriv
    - - ~
      - Primitive:
          Interception: Operation
    - - ~
      - Composite:
          name: AuthContext
          kind: Context
          fields:
            - name: id
              typ:
                Plain:
                  - String
                  - []
                  - true
              annotations:
                annotations:
                  jwt:
                    name: jwt
                    params:
                      Single:
                        - StringLiteral:
                            - sub
              default_value: ~
            - name: role
              typ:
                Plain:
                  - String
                  - []
                  - true
              annotations:
                annotations:
                  jwt:
                    name: jwt
                    params: None
              default_value: ~
          annotations:
            annotations: {}
    - - ~
      - Composite:
          name: Document
          kind: Type
          fields:
            - name: id
              typ:
                Plain:
                  - Int
                  - []
                  - true
              annotations:
                annotations:
                  pk:
                    name: pk
                    params: None
              default_value:
                kind:
                  Function:
                    - autoIncrement
                    - []
            - name: content
              typ:
                Plain:
                  - String
                  - []
                  - true
              annotations:
                annotations: {}
              default_value: ~
            - name: documentUsers
              typ:
                Plain:
                  - Set
                  - - Plain:
                        - DocumentUser
                        - []
                        - true
                  - true
              annotations:
                annotations: {}
              default_value: ~
          annotations:
            annotations:
              access:
                name: access
                params:
                  Map:
                    - mutation:
                        FieldSelection:
                          Select:
                            - Select:
                                - Single:
                                    - Identifier:
                                        - self
                                        - Reference:
                                            index: 17
                                            generation: ~
                                    - Reference:
                                        index: 17
                                        generation: ~
                                - Identifier:
                                    - documentUsers
                                    - Defer
                                - Set:
                                    Reference:
                                      index: 18
                                      generation: ~
                            - HofCall:
                                name:
                                  - some
                                param_name:
                                  - du
                                expr:
                                  LogicalOp:
                                    And:
                                      - LogicalOp:
                                          Or:
                                            - RelationalOp:
                                                Eq:
                                                  - FieldSelection:
                                                      Select:
                                                        - Single:
                                                            - Identifier:
                                                                - AuthContext
                                                                - Reference:
                                                                    index: 16
                                                                    generation: ~
                                                            - Reference:
                                                                index: 16
                                                                generation: ~
                                                        - Identifier:
                                                            - role
                                                            - Defer
                                                        - Reference:
                                                            index: 4
                                                            generation: ~
                                                  - StringLiteral:
                                                      - admin
                                                  - Primitive: Boolean
                                            - RelationalOp:
                                                Eq:
                                                  - FieldSelection:
                                                      Select:
                                                        - Single:
                                                            - Identifier:
                                                                - du
                                                                - Reference:
                                                                    index: 18
                                                                    generation: ~
                                                            - Reference:
                                                                index: 18
                                                                generation: ~
                                                        - Identifier:
                                                            - userId
                                                            - Defer
                                                        - Reference:
                                                            index: 4
                                                            generation: ~
                                                  - FieldSelection:
                                                      Select:
                                                        - Single:
                                                            - Identifier:
                                                                - AuthContext
                                                                - Reference:
                                                                    index: 16
                                                                    generation: ~
                                                            - Reference:
                                                                index: 16
                                                                generation: ~
                                                        - Identifier:
                                                            - id
                                                            - Defer
                                                        - Reference:
                                                            index: 4
                                                            generation: ~
                                                  - Primitive: Boolean
                                            - Primitive: Boolean
                                      - FieldSelection:
                                          Select:
                                            - Single:
                                                - Identifier:
                                                    - du
                                                    - Reference:
                                                        index: 18
                                                        generation: ~
                                                - Reference:
                                                    index: 18
                                                    generation: ~
                                            - Identifier:
                                                - write
                                                - Defer
                                            - Reference:
                                                index: 0
                                                generation: ~
                                      - Primitive: Boolean
                                typ:
                                  Primitive: Boolean
                            - Primitive: Boolean
                      query:
                        FieldSelection:
                          Select:
                            - Select:
                                - Single:
                                    - Identifier:
                                        - self
                                        - Reference:
                                            index: 17
                                            generation: ~
                                    - Reference:
                                        index: 17
                                        generation: ~
                                - Identifier:
                                    - documentUsers
                                    - Defer
                                - Set:
                                    Reference:
                                      index: 18
                                      generation: ~
                            - HofCall:
                                name:
                                  - some
                                param_name:
                                  - du
                                expr:
                                  LogicalOp:
                                    And:
                                      - LogicalOp:
                                          Or:
                                            - RelationalOp:
                                                Eq:
                                                  - FieldSelection:
                                                      Select:
                                                        - Single:
                                                            - Identifier:
                                                                - AuthContext
                                                                - Reference:
                                                                    index: 16
                                                                    generation: ~
                                                            - Reference:
                                                                index: 16
                                                                generation: ~
                                                        - Identifier:
                                                            - role
                                                            - Defer
                                                        - Reference:
                                                            index: 4
                                                            generation: ~
                                                  - StringLiteral:
                                                      - admin
                                                  - Primitive: Boolean
                                            - RelationalOp:
                                                Eq:
                                                  - FieldSelection:
                                                      Select:
                                                        - Single:
                                                            - Identifier:
                                                                - du
                                                                - Reference:
                                                                    index: 18
                                                                    generation: ~
                                                            - Reference:
                                                                index: 18
                                                                generation: ~
                                                        - Identifier:
                                                            - userId
                                                            - Defer
                                                        - Reference:
                                                            index: 4
                                                            generation: ~
                                                  - FieldSelection:
                                                      Select:
                                                        - Single:
                                                            - Identifier:
                                                                - AuthContext
                                                                - Reference:
                                                                    index: 16
                                                                    generation: ~
                                                            - Reference:
                                                                index: 16
                                                                generation: ~
                                                        - Identifier:
                                                            - id
                                                            - Defer
                                                        - Reference:
                                                            index: 4
                                                            generation: ~
                                                  - Primitive: Boolean
                                            - Primitive: Boolean
                                      - FieldSelection:
                                          Select:
                                            - Single:
                                                - Identifier:
                                                    - du
                                                    - Reference:
                                                        index: 18
                                                        generation: ~
                                                - Reference:
                                                    index: 18
                                                    generation: ~
                                            - Identifier:
                                                - read
                                                - Defer
                                            - Reference:
                                                index: 0
                                                generation: ~
                                      - Primitive: Boolean
                                typ:
                                  Primitive: Boolean
                            - Primitive: Boolean
    - - ~
      - Composite:
          name: DocumentUser
          kind: Type
          fields:
            - name: id
              typ:
                Plain:
                  - Int
                  - []
                  - true
              annotations:
                annotations:
                  pk:
                    name: pk
                    params: None
              default_value:
                kind:
                  Function:
                    - autoIncrement
                    - []
            - name: document
              typ:
                Plain:
                  - Document
                  - []
                  - true
              annotations:
                annotations: {}
              default_value: ~
            - name: userId
              typ:
                Plain:
                  - String
                  - []
                  - true
              annotations:
                annotations: {}
              default_value: ~
            - name: read
              typ:
                Plain:
                  - Boolean
                  - []
                  - true
              annotations:
                annotations: {}
              default_value: ~
            - name: write
              typ:
                Plain:
                  - Boolean
                  - []
                  - true
              annotations:
                annotations: {}
              default_value: ~
          annotations:
            annotations: {}
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
  map:
    AuthContext:
      index: 16
      generation: ~
    Blob:
      index: 10
      generation: ~
    Boolean:
      index: 0
      generation: ~
    Decimal:
      index: 3
      generation: ~
    Document:
      index: 17
      generation: ~
    DocumentUser:
      index: 18
      generation: ~
    Exograph:
      index: 13
      generation: ~
    ExographPriv:
      index: 14
      generation: ~
    Float:
      index: 2
      generation: ~
    Instant:
      index: 8
      generation: ~
    Int:
      index: 1
      generation: ~
    Json:
      index: 9
      generation: ~
    LocalDate:
      index: 7
      generation: ~
    LocalDateTime:
      index: 6
      generation: ~
    LocalTime:
      index: 5
      generation: ~
    Operation:
      index: 15
      generation: ~
    String:
      index: 4
      generation: ~
    Uuid:
      index: 11
      generation: ~
    Vector:
      index: 12
      generation: ~
modules:
  values:
    - - ~
      - name: DocsDatabase
        annotations:
          annotations:
            postgres:
              name: postgres
              params: None
        types:
          - name: Document
            kind: Type
            fields:
              - name: id
                typ:
                  Plain:
                    - Int
                    - []
                    - true
                annotations:
                  annotations:
                    pk:
                      name: pk
                      params: None
                default_value:
                  kind:
                    Function:
                      - autoIncrement
                      - []
              - name: content
                typ:
                  Plain:
                    - String
                    - []
                    - true
                annotations:
                  annotations: {}
                default_value: ~
              - name: documentUsers
                typ:
                  Plain:
                    - Set
                    - - Plain:
                          - DocumentUser
                          - []
                          - true
                    - true
                annotations:
                  annotations: {}
                default_value: ~
            annotations:
              annotations:
                access:
                  name: access
                  params:
                    Map:
                      - mutation:
                          FieldSelection:
                            Select:
                              - Select:
                                  - Single:
                                      - Identifier:
                                          - self
                                          - Reference:
                                              index: 17
                                              generation: ~
                                      - Reference:
                                          index: 17
                                          generation: ~
                                  - Identifier:
                                      - documentUsers
                                      - Defer
                                  - Set:
                                      Reference:
                                        index: 18
                                        generation: ~
                              - HofCall:
                                  name:
                                    - some
                                  param_name:
                                    - du
                                  expr:
                                    LogicalOp:
                                      And:
                                        - LogicalOp:
                                            Or:
                                              - RelationalOp:
                                                  Eq:
                                                    - FieldSelection:
                                                        Select:
                                                          - Single:
                                                              - Identifier:
                                                                  - AuthContext
                                                                  - Reference:
                                                                      index: 16
                                                                      generation: ~
                                                              - Reference:
                                                                  index: 16
                                                                  generation: ~
                                                          - Identifier:
                                                              - role
                                                              - Defer
                                                          - Reference:
                                                              index: 4
                                                              generation: ~
                                                    - StringLiteral:
                                                        - admin
                                                    - Primitive: Boolean
                                              - RelationalOp:
                                                  Eq:
                                                    - FieldSelection:
                                                        Select:
                                                          - Single:
                                                              - Identifier:
                                                                  - du
                                                                  - Reference:
                                                                      index: 18
                                                                      generation: ~
                                                              - Reference:
                                                                  index: 18
                                                                  generation: ~
                                                          - Identifier:
                                                              - userId
                                                              - Defer
                                                          - Reference:
                                                              index: 4
                                                              generation: ~
                                                    - FieldSelection:
                                                        Select:
                                                          - Single:
                                                              - Identifier:
                                                                  - AuthContext
                                                                  - Reference:
                                                                      index: 16
                                                                      generation: ~
                                                              - Reference:
                                                                  index: 16
                                                                  generation: ~
                                                          - Identifier:
                                                              - id
                                                              - Defer
                                                          - Reference:
                                                              index: 4
                                                              generation: ~
                                                    - Primitive: Boolean
                                              - Primitive: Boolean
                                        - FieldSelection:
                                            Select:
                                              - Single:
                                                  - Identifier:
                                                      - du
                                                      - Reference:
                                                          index: 18
                                                          generation: ~
                                                  - Reference:
                                                      index: 18
                                                      generation: ~
                                              - Identifier:
                                                  - write
                                                  - Defer
                                              - Reference:
                                                  index: 0
                                                  generation: ~
                                        - Primitive: Boolean
                                  typ:
                                    Primitive: Boolean
                              - Primitive: Boolean
                        query:
                          FieldSelection:
                            Select:
                              - Select:
                                  - Single:
                                      - Identifier:
                                          - self
                                          - Reference:
                                              index: 17
                                              generation: ~
                                      - Reference:
                                          index: 17
                                          generation: ~
                                  - Identifier:
                                      - documentUsers
                                      - Defer
                                  - Set:
                                      Reference:
                                        index: 18
                                        generation: ~
                              - HofCall:
                                  name:
                                    - some
                                  param_name:
                                    - du
                                  expr:
                                    LogicalOp:
                                      And:
                                        - LogicalOp:
                                            Or:
                                              - RelationalOp:
                                                  Eq:
                                                    - FieldSelection:
                                                        Select:
                                                          - Single:
                                                              - Identifier:
                                                                  - AuthContext
                                                                  - Reference:
                                                                      index: 16
                                                                      generation: ~
                                                              - Reference:
                                                                  index: 16
                                                                  generation: ~
                                                          - Identifier:
                                                              - role
                                                              - Defer
                                                          - Reference:
                                                              index: 4
                                                              generation: ~
                                                    - StringLiteral:
                                                        - admin
                                                    - Primitive: Boolean
                                              - RelationalOp:
                                                  Eq:
                                                    - FieldSelection:
                                                        Select:
                                                          - Single:
                                                              - Identifier:
                                                                  - du
                                                                  - Reference:
                                                                      index: 18
                                                                      generation: ~
                                                              - Reference:
                                                                  index: 18
                                                                  generation: ~
                                                          - Identifier:
                                                              - userId
                                                              - Defer
                                                          - Reference:
                                                              index: 4
                                                              generation: ~
                                                    - FieldSelection:
                                                        Select:
                                                          - Single:
                                                              - Identifier:
                                                                  - AuthContext
                                                                  - Reference:
                                                                      index: 16
                                                                      generation: ~
                                                              - Reference:
                                                                  index: 16
                                                                  generation: ~
                                                          - Identifier:
                                                              - id
                                                              - Defer
                                                          - Reference:
                                                              index: 4
                                                              generation: ~
                                                    - Primitive: Boolean
                                              - Primitive: Boolean
                                        - FieldSelection:
                                            Select:
                                              - Single:
                                                  - Identifier:
                                                      - du
                                                      - Reference:
                                                          index: 18
                                                          generation: ~
                                                  - Reference:
                                                      index: 18
                                                      generation: ~
                                              - Identifier:
                                                  - read
                                                  - Defer
                                              - Reference:
                                                  index: 0
                                                  generation: ~
                                        - Primitive: Boolean
                                  typ:
                                    Primitive: Boolean
                              - Primitive: Boolean
          - name: DocumentUser
            kind: Type
            fields:
              - name: id
                typ:
                  Plain:
                    - Int
                    - []
                    - true
                annotations:
                  annotations:
                    pk:
                      name: pk
                      params: None
                default_value:
                  kind:
                    Function:
                      - autoIncrement
                      - []
              - name: document
                typ:
                  Plain:
                    - Document
                    - []
                    - true
                annotations:
                  annotations: {}
                default_value: ~
              - name: userId
                typ:
                  Plain:
                    - String
                    - []
                    - true
                annotations:
                  annotations: {}
                default_value: ~
              - name: read
                typ:
                  Plain:
                    - Boolean
                    - []
                    - true
                annotations:
                  annotations: {}
                default_value: ~
              - name: write
                typ:
                  Plain:
                    - Boolean
                    - []
                    - true
                annotations:
                  annotations: {}
                default_value: ~
            annotations:
              annotations: {}
        methods: []
        interceptors: []
        base_exofile: input.exo
    - ~
    - ~
    - ~
  map:
    DocsDatabase:
      index: 0
      generation: ~
